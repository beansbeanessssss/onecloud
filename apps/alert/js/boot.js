define({

	 /**
 	 * allow to setup the application
 	 *
 	 * @return promise
	 * @param self will feed back 'this' object
 	 */

	setup: () => {
		var p = new Promise((r, d) => {
			requirejs([OC.appJS('alert', 'alert.bundle')], () => {
				r({
					id: 'alert',
					start : true
				});
			});
		});
		return p;
	},

	boot: () => { return false }
});
