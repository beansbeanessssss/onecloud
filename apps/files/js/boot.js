define({

	 	/**
		 * allow to setup the application
		 *
		 * @return promise
		 * @param self will feed back 'this' object
		 */

	 	setup: () => {
	 		var my = self.info;
	 		var p = new Promise((resolve, defer) => {
	 			OC.registerNavItem('files', {
	 				name: 'Files',
	 				iconMaterial: 'folder',
	 				route: '/'
	 			});
				resolve( { id : 'files' } );
	 		});
	 		return p;
	 	},


		/**
		 * Start the application by mounting it to the respective DOM element
		 *
		 * @return promise
		 */

		boot: (container) => {
			var p = new Promise((resolve, defer) => {
				requirejs([OC.appJS('files', 'files.bundle')], (app) => {
					app.$mount(container);
					app.$once('mounted', resolve( { start: true } ) );
				});
			});
			return p;
		}
});
