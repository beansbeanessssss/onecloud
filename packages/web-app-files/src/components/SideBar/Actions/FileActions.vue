<template>
  <oc-list id="oc-files-actions-sidebar" class="oc-mt-s">
    <action-menu-item
      v-for="(action, index) in actions"
      :key="`action-${index}`"
      :action="action"
      :items="resources"
      class="oc-py-xs"
    />
  </oc-list>
</template>

<script>
import { mapGetters } from 'vuex'
import ActionMenuItem from '../../ActionMenuItem.vue'

import FileActions from '../../../mixins/fileActions'
import { useIsFilesAppActive } from '../../../composables/useIsFilesAppActive'

export default {
  name: 'FileActions',
  components: { ActionMenuItem },
  mixins: [FileActions],
  setup() {
    return {
      isFilesAppActive: useIsFilesAppActive()
    }
  },
  title: ($gettext) => {
    return $gettext('Actions')
  },
  computed: {
    ...mapGetters('Files', ['highlightedFile']),

    resources() {
      return [this.highlightedFile]
    },

    actions() {
      return this.$_fileActions_getAllAvailableActions(this.resources)
    }
  }
}
</script>

<style lang="scss">
#oc-files-actions-sidebar {
  > li a,
  > li a:hover {
    color: var(--oc-color-swatch-passive-default);
    display: inline-flex;
    gap: 10px;
    vertical-align: top;
  }
}
</style>
