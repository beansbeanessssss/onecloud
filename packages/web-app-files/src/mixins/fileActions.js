import get from 'lodash-es/get'
import { mapGetters, mapActions, mapState } from 'vuex'

import { checkRoute } from '../helpers/route'
import AcceptShare from './actions/acceptShare'
import Copy from './actions/copy'
import DeclineShare from './actions/declineShare'
import Delete from './actions/delete'
import DownloadFile from './actions/downloadFile'
import DownloadFolder from './actions/downloadFolder'
import Favorite from './actions/favorite'
import Fetch from './actions/fetch'
import Move from './actions/move'
import Navigate from './actions/navigate'
import Rename from './actions/rename'
import Restore from './actions/restore'
import kebabCase from 'lodash-es/kebabCase'

const actionsMixins = [
  'fetch',
  'navigate',
  'downloadFile',
  'downloadFolder',
  'favorite',
  'copy',
  'move',
  'rename',
  'restore',
  'delete',
  'acceptShare',
  'declineShare'
]

export const EDITOR_MODE_EDIT = 'edit'
export const EDITOR_MODE_CREATE = 'create'

export default {
  mixins: [
    AcceptShare,
    Copy,
    DeclineShare,
    Delete,
    DownloadFile,
    DownloadFolder,
    Favorite,
    Fetch,
    Move,
    Navigate,
    Rename,
    Restore
  ],
  computed: {
    ...mapState(['apps']),
    ...mapGetters('Files', ['highlightedFile', 'currentFolder']),
    ...mapGetters('External', ['mimeTypes']),
    ...mapGetters(['capabilities', 'configuration']),

    $_fileActions_systemActions() {
      const systemActions = []

      for (const actionMixin of actionsMixins) {
        systemActions.push(...this[`$_${actionMixin}_items`])
      }

      return systemActions
    },

    $_fileActions_editorActions() {
      return this.apps.fileEditors.map((editor) => {
        return {
          label: () => {
            const translated = this.$gettext('Open in %{app}')
            return this.$gettextInterpolate(
              translated,
              { app: this.apps.meta[editor.app].name },
              true
            )
          },
          icon: this.apps.meta[editor.app].icon,
          handler: (item) =>
            this.$_fileActions_openEditor(editor, item.path, item.id, EDITOR_MODE_EDIT),
          isEnabled: ({ resource }) => {
            if (editor.routes?.length > 0 && !checkRoute(editor.routes, this.$route.name)) {
              return false
            }

            return resource.extension === editor.extension
          },
          canBeDefault: true,
          componentType: 'oc-button',
          class: `oc-files-actions-${kebabCase(
            this.apps.meta[editor.app].name
          ).toLowerCase()}-trigger`
        }
      })
    }
  },

  methods: {
    ...mapActions(['openFile']),

    $_fileActions_openEditor(editor, filePath, fileId, mode) {
      if (editor.handler) {
        return editor.handler({
          config: this.configuration,
          extensionConfig: editor.config,
          filePath,
          fileId,
          mode
        })
      }

      // TODO: Refactor in the store
      this.openFile({
        filePath
      })

      if (editor.newTab) {
        const path = this.$router.resolve({
          name: editor.routeName,
          params: { filePath, fileId, mode }
        }).href
        const target = `${editor.routeName}-${filePath}`
        const win = window.open(path, target)
        // in case popup is blocked win will be null
        if (win) {
          win.focus()
        }
        return
      }

      this.$router.push({
        name: editor.routeName || editor.app,
        params: {
          filePath,
          fileId,
          mode,
          contextRouteName: this.$route.name
        }
      })
    },

    // TODO: Make user-configurable what is a defaultAction for a filetype/mimetype
    // returns the _first_ action from actions array which we now construct from
    // available mime-types coming from the app-provider and existing actions
    $_fileActions_triggerDefaultAction(resource) {
      const actions = this.$_fileActions_getAllAvailableActions(resource).filter(
        (action) => action.canBeDefault
      )
      actions[0].handler(resource, actions[0].handlerData)
    },

    $_fileActions_getAllAvailableActions(resource) {
      return [
        ...this.$_fileActions_editorActions,
        ...this.$_fileActions_loadExternalAppActions(resource),
        ...this.$_fileActions_systemActions
      ].filter((action) => {
        return action.isEnabled({
          resource,
          parent: this.currentFolder
        })
      })
    },

    // returns an array of available external Apps
    // to open a resource with a specific mimeType
    // FIXME: filesApp should not know anything about any other app, dont cross the line!!! BAD
    $_fileActions_loadExternalAppActions(resource) {
      const { mimeType } = resource
      if (
        mimeType === undefined ||
        !get(this, 'capabilities.files.app_providers') ||
        !get(this, 'mimeTypes', []).length
      ) {
        return []
      }

      const { app_providers: appProviders = [] } = this.mimeTypes.find(
        (t) => t.mime_type === mimeType
      )

      return appProviders.map((app) => {
        const label = this.$gettext('Open in %{ appName }')
        return {
          img: app.icon,
          componentType: 'oc-button',
          class: `oc-files-actions-${app.name}-trigger`,
          isEnabled: () => true,
          canBeDefault: true,
          handler: () => {
            const routeData = this.$router.resolve({
              name: 'external-apps',
              params: { app: app.name, file_id: resource.fileId },
              // public-token retrieval is weak, same as packages/web-app-files/src/index.js:106
              query: {
                ...(this.isPublicPage && {
                  'public-token': (this.$route.params.item || '').split('/')[0]
                })
              }
            })
            // TODO: Let users configure whether to open in same/new tab (`_blank` vs `_self`)
            window.open(routeData.href, '_blank')
          },
          label: () => this.$gettextInterpolate(label, { appName: app.name })
        }
      })
    }
  }
}
