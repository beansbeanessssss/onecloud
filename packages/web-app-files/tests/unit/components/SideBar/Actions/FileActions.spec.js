import Vuex from 'vuex'
import DesignSystem from 'owncloud-design-system'
import stubs from '@/tests/unit/stubs/index.js'
import { createLocalVue, mount } from '@vue/test-utils'
import FileActions from '@files/src/components/SideBar/Actions/FileActions.vue'

import GetTextPlugin from 'vue-gettext'

import { apps, getActions, fileActions } from '@files/tests/__fixtures__/fileActions.js'
import fixtureMimeTypes from '@files/tests/__fixtures__/mimeTypes.js'

const localVue = createLocalVue()
localVue.use(DesignSystem)
localVue.use(Vuex)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

const filesPersonalRoute = { name: 'files-personal' }

const systemDefaultActions = [
  [['copy']],
  [['copy', 'move']],
  [['copy', 'markdown-editor']],
  [['copy', 'move', 'download', 'markdown-editor']]
]

const exampleApps = [
  { name: 'exampleName1', icon: 'exampleIcon1' },
  { name: 'exampleName2', icon: 'exampleIcon2' },
  { name: 'exampleName3', icon: 'exampleIcon3' }
]

describe('FileActions', () => {
  describe('when user is on personal route', () => {
    afterEach(() => {
      jest.clearAllMocks()
    })

    describe('and there are no externalApps', () => {
      it.each(systemDefaultActions)('only default actions should be rendered', actions => {
        const wrapper = getWrapper(filesPersonalRoute, actions)

        expect(wrapper).toMatchSnapshot()
      })
    })

    describe('and there are externalApps', () => {
      it.each(systemDefaultActions)('default and external actions should be rendered', actions => {
        const wrapper = getWrapper(
          filesPersonalRoute,
          actions,
          createStore({
            mimeType: 'application/fileFormat3',
            availableMimeTypes: fixtureMimeTypes
          }),
          exampleApps
        )

        expect(wrapper).toMatchSnapshot()
      })
    })

    describe('should trigger the action handlers on click', () => {
      it('for external App actions', async () => {
        const triggerActionSpy = jest
          .spyOn(FileActions.mixins[0].methods, '$_fileActions_openLink')
          .mockImplementationOnce(jest.fn)

        const wrapper = getWrapper(
          filesPersonalRoute,
          [],
          createStore({
            mimeType: 'application/fileFormat3',
            availableMimeTypes: fixtureMimeTypes
          }),
          exampleApps
        )

        const app = exampleApps[0]
        const appButton = wrapper.find('#oc-files-actions-sidebar > li > button')
        expect(appButton.text()).toBe(`Open in ${app.name}`)

        await appButton.trigger('click')
        expect(triggerActionSpy).toHaveBeenCalledWith(app.name, 'a93f8adf==')
      })

      it('for default file and system actions', async () => {
        const actions = ['copy', 'download', 'move', 'open-folder', 'markdown-editor']
        const wrapper = getWrapper(filesPersonalRoute, actions)

        for (const button of actions) {
          const action = fileActions[button]

          const buttonElement = wrapper.find(action.selector)
          expect(buttonElement.exists()).toBeTruthy()

          await buttonElement.trigger('click.stop')
          expect(action.handler).toHaveBeenCalledTimes(1)
        }
      })
    })
  })
})

function getWrapper(
  route,
  actions = [],
  store = createStore({
    mimeType: undefined,
    availableMimeTypes: {}
  }),
  availableApps = []
) {
  const mountStubs = { ...stubs, 'oc-button': false }
  const mountOptions = {
    localVue,
    store,
    data() {
      return {
        appList: availableApps
      }
    },
    stubs: mountStubs,
    mocks: {
      $route: route,
      publicPage: () => false
    },
    computed: {
      actions: () => getActions(actions)
    }
  }
  return mount(FileActions, mountOptions)
}

function createStore(state, filename, fileId, extension, type, mimeType, availableMimeTypes) {
  return new Vuex.Store({
    state: {
      apps: apps
    },
    modules: {
      External: {
        namespaced: true,
        getters: {
          getMimeTypes: () => {
            return availableMimeTypes
          }
        }
      },
      Files: {
        state: {
          ...state,
          currentFolder: { path: '' }
        },
        namespaced: true,
        getters: {
          highlightedFile: () => {
            return getResource({ filename, fileId, extension, type, mimeType })
          },
          currentFolder: () => '/'
        }
      }
    }
  })
}

function getResource({
  filename = 'welcome',
  fileId = 'a93f8adf==',
  extension = 'md',
  type = 'file',
  mimeType
}) {
  return {
    id: '4',
    fileId,
    icon: type,
    name: type === 'file' ? `${filename}.${extension}` : filename,
    extension,
    path: type === 'file' ? `/${filename}.${extension}` : `/${filename}`,
    type,
    mdate: 'Mon, 12 Jul 2021 11:04:33 GMT',
    size: '163',
    indicators: [],
    permissions: 'RDNVW',
    starred: false,
    etag: '"89128c0e8122002db57bd19c9ec33004"',
    mimeType,
    shareTypes: [],
    downloadURL: '',
    ownerDisplayName: 'user1',
    ownerId: 'user1',
    canDownload: () => true,
    isReceivedShare: () => true,
    canBeDeleted: () => true,
    canRename: () => true
  }
}
