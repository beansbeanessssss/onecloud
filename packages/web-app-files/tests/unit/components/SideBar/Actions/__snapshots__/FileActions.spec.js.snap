// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileActions when user is on personal route should show the actions correctly 1`] = `
<ul id="oc-files-actions-sidebar" class="uk-list oc-mt-s">
  <li class="oc-py-xs"><button class="oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-text-bold oc-files-actions-sidebar-copy-trigger">
      <oc-icon-stub name="copy" accessiblelabel="" type="span" size="medium" variation="passive"></oc-icon-stub> <span class="oc-files-actions-sidebar-action-label">Copy</span>
      <!---->
    </button></li>
</ul>
`;

exports[`FileActions when user is on personal route should show the actions correctly 2`] = `
<ul id="oc-files-actions-sidebar" class="uk-list oc-mt-s">
  <li class="oc-py-xs"><button class="oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-text-bold oc-files-actions-sidebar-copy-trigger">
      <oc-icon-stub name="copy" accessiblelabel="" type="span" size="medium" variation="passive"></oc-icon-stub> <span class="oc-files-actions-sidebar-action-label">Copy</span>
      <!---->
    </button></li>
  <li class="oc-py-xs"><button class="oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-text-bold oc-files-actions-sidebar-move-trigger">
      <oc-icon-stub name="move" accessiblelabel="" type="span" size="medium" variation="passive"></oc-icon-stub> <span class="oc-files-actions-sidebar-action-label">Move</span>
      <!---->
    </button></li>
</ul>
`;

exports[`FileActions when user is on personal route should show the actions correctly 3`] = `
<ul id="oc-files-actions-sidebar" class="uk-list oc-mt-s">
  <li class="oc-py-xs"><button class="oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-text-bold oc-files-actions-sidebar-copy-trigger">
      <oc-icon-stub name="copy" accessiblelabel="" type="span" size="medium" variation="passive"></oc-icon-stub> <span class="oc-files-actions-sidebar-action-label">Copy</span>
      <!---->
    </button></li>
  <li class="oc-py-xs"><button class="oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-text-bold oc-files-actions-sidebar-markdown-editor-trigger">
      <oc-icon-stub name="markdown-editor" accessiblelabel="" type="span" size="medium" variation="passive"></oc-icon-stub> <span class="oc-files-actions-sidebar-action-label">Open in Markdown Editor</span> <span class="oc-invisible-sr">(Opens in new window)</span>
    </button></li>
</ul>
`;

exports[`FileActions when user is on personal route should show the actions correctly 4`] = `
<ul id="oc-files-actions-sidebar" class="uk-list oc-mt-s">
  <li class="oc-py-xs"><button class="oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-text-bold oc-files-actions-sidebar-copy-trigger">
      <oc-icon-stub name="copy" accessiblelabel="" type="span" size="medium" variation="passive"></oc-icon-stub> <span class="oc-files-actions-sidebar-action-label">Copy</span>
      <!---->
    </button></li>
  <li class="oc-py-xs"><button class="oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-text-bold oc-files-actions-sidebar-move-trigger">
      <oc-icon-stub name="move" accessiblelabel="" type="span" size="medium" variation="passive"></oc-icon-stub> <span class="oc-files-actions-sidebar-action-label">Move</span>
      <!---->
    </button></li>
  <li class="oc-py-xs"><button class="oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-text-bold oc-files-actions-sidebar-download-trigger">
      <oc-icon-stub name="download" accessiblelabel="" type="span" size="medium" variation="passive"></oc-icon-stub> <span class="oc-files-actions-sidebar-action-label">Download</span>
      <!---->
    </button></li>
  <li class="oc-py-xs"><button class="oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-text-bold oc-files-actions-sidebar-markdown-editor-trigger">
      <oc-icon-stub name="markdown-editor" accessiblelabel="" type="span" size="medium" variation="passive"></oc-icon-stub> <span class="oc-files-actions-sidebar-action-label">Open in Markdown Editor</span> <span class="oc-invisible-sr">(Opens in new window)</span>
    </button></li>
</ul>
`;
